aws-crt-kotlin/native/src/aws/sdk/kotlin/crt/util/hashing/ShaNative.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import aws.sdk.kotlin.crt.Allocator
+import aws.sdk.kotlin.crt.CrtRuntimeException
+import aws.sdk.kotlin.crt.awsAssertOpSuccess
+import kotlinx.cinterop.*
+import libcrt.*
+
+/**
+ * A typealias to CRT's aws_<checksum_algorithm>_new functions (i.e. aws_sha1_new, aws_sha256_new))
+ */
+internal typealias InitializeHashFn = (
+    allocator: CValuesRef<aws_allocator>?,
+) -> CPointer<aws_hash>?
+
+/**
+ * SHA-1 hash function implemented using bindings to CRT
+ */
+public class Sha1 : HashFunction {
+    private val sha1 = Sha(::aws_sha1_new)
+    override fun update(input: ByteArray, offset: Int, length: Int) { sha1.update(input, offset, length) }
+    override fun digest(): ByteArray = sha1.digest()
+    override fun reset() { sha1.reset() }
+}
+
+/**
+ * SHA-256 hash function implemented using bindings to CRT
+ */
+public class Sha256 : HashFunction {
+    private val sha256 = Sha(::aws_sha256_new)
+    override fun update(input: ByteArray, offset: Int, length: Int) { sha256.update(input, offset, length) }
+    override fun digest(): ByteArray = sha256.digest()
+    override fun reset() { sha256.reset() }
+}
+
+internal class Sha(val initializeFn: InitializeHashFn) : HashFunction {
+    private var hash: CPointer<aws_hash>
+
+    init {
+        hash = initializeHash()
+    }
+
+    // aws_hash_update
+    override fun update(input: ByteArray, offset: Int, length: Int) {
+        val inputCursor = input.usePinned {
+            aws_byte_cursor_from_array(it.addressOf(offset), length.convert())
+        }
+        awsAssertOpSuccess(aws_hash_update(hash, inputCursor)) {
+            "aws_hash_update"
+        }
+    }
+
+    // aws_hash_finalize
+    override fun digest(): ByteArray {
+        val output: CPointer<aws_byte_buf> = aws_mem_calloc(
```

Comment:
**suggestion**. You can probably do this without an allocation using `ByteArray` directly by constructing the byte buf yourself.

Something like:
```kotlin
val output = ByteArray(32)
output.usePinned { pinned ->
     val buf = cValueOf<aws_byte_buf> {
          capacity = ...
         len = 0
         buffer = addressOf(0)
      }
}
```